### ISSUE TITLE
Exhaustive return points fails for two union typed variables

### ISSUE THREAD

**Issue by ianliu** [microsoft/pyright#2139]
The following code example correctly checks that `bar` should always return a boolean, but fails for `foo` which handles nested type checks. The `reveal_type`'s calls shows that the inferred types are correct in all spots.

```python
from typing import Union

def bar(a: Union[str, int]) -> bool:
    if isinstance(a, str):
        return False
    elif isinstance(a, int):
        return False
    else:
        reveal_type(a) # Type of "a" is Never

def foo(a: Union[str, int], b: Union[str, int]) -> bool: # Type "None" cannot be assigned to type "bool"
    if isinstance(a, str):
        return True
    elif isinstance(a, int):
        if isinstance(b, str):
            return False
        elif isinstance(b, int):
            return False
        else:
            reveal_type(b) # Type of "b" is Never
    else:
        reveal_type(a) # Type of "a" is Never
```

**Comment by erictraut** [comment]
This is behaving as intended. Pyright supports a feature called ["narrowing for implied else"](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#narrowing-for-implied-else), which handles typical exhaustive return patterns. In your case, you're including the `else` clauses even though they are provably never needed. If you enable pyright's `reportUnnecessaryIsInstance` check (which is on by default in strict mode), it will highlight the fact that some of your `isinstance` calls are unnecessary. If you remove the unnecessary code, it will type check fine, and it will be robust to change in the code (e.g. if you expand the union to include more types in the future).

```python
from typing import Union

def bar(a: Union[str, int]) -> bool:
    if isinstance(a, str):
        return False
    else:
        return False

def foo(a: Union[str, int], b: Union[str, int]) -> bool:
    if isinstance(a, str):
        return True
    else:
        if isinstance(b, str):
            return False
        else:
            return False
```

I will also explore adding logic to handle the exhaustive return usage pattern in your code sample above. I'll see if I can find a way to add it without introducing too much performance overhead or false positives.

**Comment by ianliu** [comment]
Sorry, I think I couldn't understand your point. If I remove the else's clauses, I still get the error. Shouldn't this type-check correctly?

```python
def bar(a: Union[str, int]) -> bool:
    if isinstance(a, str):
        return False
    elif isinstance(a, int):
        return False

def foo(a: Union[str, int], b: Union[str, int]) -> bool:
    if isinstance(a, str):
        return True
    elif isinstance(a, int):
        if isinstance(b, str):
            return False
        elif isinstance(b, int):
            return False
```
Here I'm returning a boolean in all possible cases of `a` and `b`, but pyright still warns me about `foo` with: `Type "None" cannot be assigned to type "bool"`

**Comment by ianliu** [comment]
Just to add as a motivation: I would like to keep `elif`'s in case the type changes in the future. That way I would be warned by the type checker where I need to fix my code. If I were to use an `else` clause, I wouldn't have this benefit.

This is also stated in the documentation you've linked above:

> If you later added another color to the Color enumeration above (e.g. YELLOW = 4), Pyright would detect that func3 no longer exhausts all members of the enumeration and possibly returns None

**Comment by erictraut** [comment]
The "implied elif" logic works with a single variable but not with multiple variables. That's why your first example `bar` works when you remove the `else` statements but the second does not.

Yes, I agree there's value in retaining the `elif` statements for the reason you noted.

**Comment by erictraut** [comment]
The challenge here is that pyright's type analysis engine is a just-in-time evaluator. When it is asked to evaluate the type of a parse tree node, it can satisfy that request by evaluating a minimal number of other dependent types. It hops around (even between files) as necessary to satisfy the original request, and it caches the type of evaluated nodes in memory to satisfy subsequent requests. 

By contrast, mypy always analyzes entire files, and it performs multiple passes until all types in the file converge.

Pyright's approach performs much better, and it is key to providing interactive language service features like completion suggestions, which must return results to the user with minimal delay.

Pyright is able to do a just-in-time analysis by building the control flow graph before evaluating types. It doesn't perform multiple passes, so it cannot discover changes to the code flow graph based on new type analysis information. The code flow graph necessarily needs to be built first. 

In the code sample above, determining whether the `else` portion of the code flow graph can be executed requires the evaluation of types. In other words, the evaluation of types affects the code flow graph in this instance. That makes this feature very difficult to implement given pyright's architecture.

I'll continue to think about possible solutions, but I wanted to share with you some details about why this isn't a straightforward enhancement request.

**Comment by ianliu** [comment]
Thanks for the explanation. For now, I'm using an workaround like so:

```python
from typing import Union

def bar(b: Union[int, str]) -> bool:
    if isinstance(b, int):
        return True
    elif isinstance(b, str):
        return True

def foo(a: Union[int, str], b: Union[int, str]) -> bool:
    if isinstance(a, int):
        return True
    elif isinstance(a, str):
        return bar(b)
```

**Comment by erictraut** [comment]
I explored adding support for exhaustive return detection that would allow your first example to work without errors. Even with some limitations, this functionality is quite expensive. It slowed down analysis by 5-7% on some of our internal benchmarks. That's a really significant slowdown that would negatively affect all pyright and pylance users. I don't think we can justify that much of a performance hit for a feature that will benefit relatively few users.

**Comment by ianliu** [comment]
Would an on/off flag be reasonable?

**Comment by ianliu** [comment]
I see two uses of pyright: on the fly check, which demands performance, and another precommit check, which doesn't depend on performance much, but catch more errors. Don't know if this makes sense :p

**Comment by erictraut** [comment]
I'll discuss this with my colleagues and get back to you.

In general, we don't like to add on/off flags for features like this because it significantly affects the behavior of analyzer (the order in which it visits nodes, etc.) and could have other side effects other than performance. If it's an optional feature that is off by default, it would be used by almost no one. That means it would be easy for a change to introduce a regression because we'd have almost no one running it against their code. Plus, it would then make issues more difficult to diagnose because we'd need to ask every person who reports a bug whether they were using this mode.

We also like to get a signal from users that a feature is desired or useful to a reasonably large subset of the user base before we add it. So far, you are the first and only person to request this feature. We would want to see a large number of upvotes (thumbs-up reactions) on this issue.

**Comment by erictraut** [comment]
Here's what we've decided. 

First, I've improved the existing "implied else" logic to handle nested cases and multiple variables. That means both of these cases will type check without errors. This fix will be in the next release of pyright and pylance.

```python
from typing import Union

def bar(a: Union[str, int]) -> bool:
    if isinstance(a, str):
        return False
    elif isinstance(a, int):
        return False

def foo(a: Union[str, int], b: Union[str, int]) -> bool:
    if isinstance(a, str):
        return True
    elif isinstance(a, int):
        if isinstance(b, str):
            return False
        elif isinstance(b, int):
            return False
```

These patterns make your code robust against changes in the future. If you were to expand the union types to include additional subtypes, the type checker would trigger an error. For example, if I extend the union for parameter `a` to include `complex`:

```python
def bar(a: Union[str, int, complex]) -> bool: # Error: Function with declared type of "bool" must return value
    if isinstance(a, str):
        return False
    elif isinstance(a, int):
        return False
```

We decided not to extend the logic to handle the explicit `else` case as in your original example. There's no additional value that this provides from a static type checking perspective, and it results in a 5-7% analysis slowdown for everyone. That's not a good tradeoff.

If you don't want to rely entirely on static checks and want to provide an `else` clause to catch the errant condition at runtime, you can raise an exception in the `else` block, and that will also type check fine.

```python
def bar(a: Union[str, int]) -> bool:
    if isinstance(a, str):
        return False
    elif isinstance(a, int):
        return False
    else:
        raise ValueError("Input value must be str or int")
```

I'll also note that Python 3.10 introduces a new and improved way to write code like this — that `match` statement. Pyright already has full support for this new language feature, including exhaustion detection.

I realize that it's too early for most people to write code that depends on Python 3.10. (It's still in beta.) But here's how the code would look using this feature:

```python
from typing import Union

def bar(a: Union[str, int]) -> bool:
    match a:
        case str():
            return False
        case int():
            return False

def foo(a: Union[str, int], b: Union[str, int]) -> bool:
    match a:
        case str():
            return True
        case int():
            match b:
                case str():
                    return False
                case int():
                    return False
```

As with the `if`/`elif` approach, this is also safe in that it properly handles exhaustion, making it robust to type changes. For example, if you expand the unions in the above example, pyright will generate an error.

**Comment by ianliu** [comment]
@erictraut that's exactly what I've expected. My original example only included the `else` clause to prove that there were no more types available in the if/elif statements; I was never expecting to have the else clause!

Thanks very much for the change.

Ian L.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.162, which I just published. It will also be in the next release of pylance.
